openzl-crypto/src/merkle_tree/partial.rs

```diff
@@ -0,0 +1,347 @@
+//! Partial Merkle Tree Storage
+
+// TODO: Do we allow custom sentinel sources for this tree?
+
+use crate::merkle_tree::{
+    capacity,
+    inner_tree::{BTreeMap, InnerMap, PartialInnerTree},
+    Configuration, CurrentPath, InnerDigest, Leaf, LeafDigest, MerkleTree, Node, Parameters, Path,
+    PathError, Root, Tree, WithProofs,
+};
+use alloc::vec::Vec;
+use core::{fmt::Debug, hash::Hash};
+use openzl_util::derivative;
+
+#[cfg(feature = "serde")]
+use openzl_util::serde::{Deserialize, Serialize};
+
+/// Partial Merkle Tree Type
+pub type PartialMerkleTree<C, M = BTreeMap<C>> = MerkleTree<C, Partial<C, M>>;
+
+/// Partial Merkle Tree Backing Structure
+#[cfg_attr(
+    feature = "serde",
+    derive(Deserialize, Serialize),
+    serde(
+        bound(
+            deserialize = "LeafDigest<C>: Deserialize<'de>, InnerDigest<C>: Deserialize<'de>, M: Deserialize<'de>",
+            serialize = "LeafDigest<C>: Serialize, InnerDigest<C>: Serialize, M: Serialize"
+        ),
+        crate = "openzl_util::serde",
+        deny_unknown_fields
+    )
+)]
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = "LeafDigest<C>: Clone, InnerDigest<C>: Clone, M: Clone"),
+    Debug(bound = "LeafDigest<C>: Debug, InnerDigest<C>: Debug, M: Debug"),
+    Default(bound = "LeafDigest<C>: Default, InnerDigest<C>: Default, M: Default"),
+    Eq(bound = "LeafDigest<C>: Eq, InnerDigest<C>: Eq, M: Eq"),
+    Hash(bound = "LeafDigest<C>: Hash, InnerDigest<C>: Hash, M: Hash"),
+    PartialEq(bound = "LeafDigest<C>: PartialEq, InnerDigest<C>: PartialEq, M: PartialEq")
+)]
+pub struct Partial<C, M = BTreeMap<C>>
+where
+    C: Configuration + ?Sized,
+    M: InnerMap<C>,
+{
+    /// Leaf Digests
+    leaf_digests: Vec<LeafDigest<C>>,
+
+    /// Inner Digests
+    inner_digests: PartialInnerTree<C, M>,
+}
```

Comment:
What is the implicit assumption on how `leaf_digests` relate to `inner_digests`?